### DISCUSSION TITLE
what are the progress so far?

### DISCUSSION THREAD

**Discussion by ThatRetroGuy** [skjelten/emusc#54] [General]
Has the sound quality been imrpoved in the code so far or is it still that same sound quality as in the pree-colmpiled build?

**Comment by skjelten** [comment]
There has only been one significant improvement since the last release, and that is the DPCM decoding that was done by @mmontag. He also expressed interest in making a go at sample interpolation, but I don't know if he have had any time to look at it yet.

For my self I have spent most of my time with writing documentation lately, but yesterday I tried nuked SC55, and just as I had hoped, it turned out to be an extremely powerful "debug tool" for EmuSC. Just by a few hours I was able to implement correct static filtering, and by only 15 minutes earlier today I was able to identify a "Level (volume) key follow" byte in the partial definition that was previously unknown. So I am very optimistic about being able to make EmuSC sound much better going forward - and that a lot faster than previously expected.

But to answer your actual question, I do not think you will hear much of a difference with current GIT vs v0.1.0 if you play regular MIDI files, but if you previously was annoyed by popping noise in looping partials then it might be worth a look.

**Reply by ThatRetroGuy** [reply]
Did wonder if you would use nuked sc55 in some way with how good that emulation is so far and since he is doing the cpu in it.
and werent too annoyed on the popping and crunchy sound start, were mostly curious in general if any sound progress had went on.
since had seen a few treads talking about sound improving. 
will there be an virtual output  from the emulator? so people can record the audio for streams as example.

**Reply by Almondatchy** [reply]
I think that the Nuked code can be used for reference and Reverse engineering the Sc-55 for the EmuSC implementation, As far as I understand, Nuked Runs a Virtual SC-55 which is limited by emulaton and what the hardware was originally programmed to do, EmuSC is a software implementation which can bypass polyphony limits and edit variables for example

**Reply by skjelten** [reply]
> will there be an virtual output from the emulator? so people can record the audio for streams as example.

I am not familiar with virtual outputs (Windows i presume?), but the audio output interface in EmuSC is quite easy to extend, so I expect that to be a relatively simple thing to do.

**Reply by ThatRetroGuy** [reply]
Yeah, a windows feature were curious on, a sort of way to record from the emusc. in the sense of saving people from routing via a virtual audio cable for example, making it an option in emusc to output on recording it.

**Comment by skjelten** [comment]
A quick update to my previous post: 
Today I was finally able to understand how the TVF (filter envelope) is working. It was surprisingly simple and needed only a handful of code lines to add support for it in EmuSC. So that went a lot faster than I had ever thought possible! The code needs some tidying up, but if I get any time this weekend I will commit a first implementation. This will improve the quality for quite a few instruments. Seems like version 0.2.0 will come sooner than expected :+1:

**Comment by Burrito78** [comment]
This is so encouraging to hear! We can't wait for EmusSC to mature to we can integrate it into DOSBox Staging! 

We tried with Nuked-SC55 but the current license prevents this.

**Reply by skjelten** [reply]
It would be a nice milestone if a project such as DOSBox Staging would want to use it :+1: 

Committed by the way support for pitch envelopes also today. Turned out I was pretty close the last time I tried, was only missing a multiplication factor. The pitch envelope seems to work pretty well now. But there are tons of refinements to be done, especially around TVF.

**Reply by ThatRetroGuy** [reply]
with these extra adjusting's, supposedly its soon with getting a V2?

**Reply by skjelten** [reply]
I think the current fixes already warrant a new version, except for one thing: I really want some kind resampler in place. Hopefully mmontag have had some time to look at it. And there was a pitch bug reported that probably should be looked into.

**Reply by ThatRetroGuy** [reply]
am very inter ested on seeing with these new fixes when it happens.

**Comment by mmontag** [comment]
I have implemented simple resampling in #55. Hopefully it's a good first step!

**Reply by ThatRetroGuy** [reply]
nice, waiting till its time for v2 to ccome out.

**Reply by skjelten** [reply]
That resampling really made a _huge_ difference in quality! So with this in place I think we are more or less ready for a new release. The only negative effect I noticed after applying the patch is that some random notes seems to be slightly delayed. It happens very seldom and it might be that my CPU is having issues keeping up with the latest interpolation - or that I have made some other bugs while hacking on the TVF and TVP envelopes.

**Reply by ThatRetroGuy** [reply]
other wise the its ready to be updated for a new release and would that maybe be soon or today?

**Reply by skjelten** [reply]
I think we might want a few days to look for bugs before a release. I just made i binary with latest git [here](https://www.skjelten.org/~skjelten/tmp/emusc/emusc_0_1_99.zip) (win32). If you have time please test it and tell us if it works well for you.

**Reply by ThatRetroGuy** [reply]
the audio is certainly an improving but there is alot of stuterring and i cant really make it output in the way i typically do with virtual audio cable, im guessing its something the program is missing, but its surely sounding worlds apart from its 0.1 release.

**Comment by skjelten** [comment]
@mmontag: If you need another challenge to enrich your life, when taking a brake from resampling, may I propose to perhaps have a look at an audio compressor? :smile:

That is the last piece of functionality that I can think of right now that must be investigated and implemented from scratch. Most other improvements onward are more of improving or extending existing implementations.

**Reply by mmontag** [reply]
What's the audio compressor?

**Reply by skjelten** [reply]
That is a function that reduces the volume gain when it goes over a certain threshold in a way that the audio never ends up clipping. Right now emusc has a tendency to clipping if you increase the master volume and reverb / chorus volumes to maximum - and play several notes at the same time. With an audio compressor you will gradually dampen the loudest output and never hear the crackling sound of clipping.

It might be that it is possible to fine tune the volume of the samples in such a way that you never need a compressor, but I have a feeling that to do so, simple instruments with no chorus / reverb will have a very low volume - compared to what you have if you play a MIDI song with many simultaneous notes and maximum system effects. 

There is absolutely a noticeable volume increase on the real hardware if you turn up the chorus and/or reverb effect when playing (so the system effects are clearly just added to the existing audio stream for that part) , but I still suspect they have some kind of limiter to make sure they never clip.

There are some information [here](https://www.uaudio.com/blog/audio-compression-basics/), but I do not think we need to boost low levels of volume, only to limit louder parts close to the legal audio range. A better link is perhaps [wikipedia](https://en.wikipedia.org/wiki/Dynamic_range_compression), which calls it dynamic range compression.

A start could perhaps be to look at munt and fluidsynth to see if they have any audio compression in their code base.

**Reply by mmontag** [reply]
Ah, okay. Yeah, I just wanted to understand whether you are talking about something built in to the SC-55 firmware, or something added on top to the software implementation. 

In my experience with fluidsynth (fluidlite) there is no fancy compressor/limiter, but only a hard clip, so it is actually easy to drive it to distortion.  

How about Nuked SC55, is it easy to make the output clip?

**Reply by skjelten** [reply]
I just tested Nuked SC55 now, playing a midi file while adjusting all part volumes and system effects to maximum and chorus feedback to an oscillating state. The audio ended up pretty loud and horrible, but never any clipping. I have never experienced clipping on the real hardware either, but perhaps never tried so hard before :slightly_smiling_face:

**Reply by MetalMaxMX** [reply]
I decided to investigate a little bit about audio compression. Is this similar to what Audacity does? https://manual.audacityteam.org/man/compressor.html? (Also has a few resources available for audio compression too)

**Reply by skjelten** [reply]
> I decided to investigate a little bit about audio compression. Is this similar to what Audacity does? https://manual.audacityteam.org/man/compressor.html? (Also has a few resources available for audio compression too)

Yes, that looks exactly like what I think we might need for emusc

**Reply by mmontag** [reply]
Here is my thinking:
- Stating the obvious, we need to fill a sample buffer with a clamped output value (i.e. [-32768,32767] or [-1,1]).
- It is possible to drive the signal to saturation/clipping. But this is a MIDI authorship/end-user problem – the user will always be able to clip the output by cranking the volume and effects to maximum, and/or play 32 loud notes at once, etc. It cannot be prevented. 
- Of course, an EmuSC user might not control MIDI authorship if they are just playing some games or old MIDI files.
- Fluidsynth's approach is to [set the default gain artificially low](https://github.com/FluidSynth/fluidsynth/issues/681#issuecomment-703766815) (let's say, 99.99% of MIDI files won't clip) and leave it up to the host application if they want to risk a little more clipping.
- A compressor/limiter would be a subjective mitigation. You might avoid hard digital clipping, but you would still have distortion.
- At this point in my line of thought, I'm tempted to say let's wait and see.
- I would love to see evidence of some software limiter in nuked sc-55, perhaps in the PCM code, before implementing it in emusc.

**Reply by skjelten** [reply]
If fluidsynth manages without a compressor I guess EmuSC should manage without it as well. 

A quick test would be to just create a worst case scenario (part and master level + reverb + chorus on all parts set to maximum - and play maximum number of polyphonic tones split on the 16 parts) and tune the sample gain so that we barley avoid clipping. My concern is that the volume for a single note without any system effect will be very low in comparison, but let us test it.

**Comment by ThatRetroGuy** [comment]


**Reply by MetalMaxMX** [reply]
Hello! As much as I believe that these roms should be in the public domain since the hardware itself has been long discontinued and I have a deep, deep seated dislike towards the ridiculous length of time that copyright has, the fact is, it is not a good idea to keep that link up. Both for you and the other contributors, I would suggest removing those links down as not doing so is outright illegal in most countries.

**Comment by farvardin** [comment]
Hello,

I don't own any read SC-55 hardware so I can't tell the difference, but the readme of emusc states: "This project is currently in development and is not able to reproduce sounds near the quality of the original synths yet". 

So far I've tried a few midifiles and my own music, and I found the sounds to be really good. 

The readme also says: "If you are looking for the best possible SC-55 emulation today you might want to try the [Nuked SC-55](https://github.com/nukeykt/Nuked-SC55) project, or to use a sound font based on the SC-55"

I can't use nuked sc-55 because on my system (linux mint 21.1 + jack) I can't make it sound correct (sound clipping and tearing). For soundfonts, I don't find they sound really good, especially in comparison to emusc.

So how far is emusc from original hardware (or from nuked sc-55)?

**Reply by skjelten** [reply]
With all the improvements we added to the 0.2.0 release I think the readme file is somewhat outdated. With the audio quality we have now the difference is really not that big anymore. Note that the actual difference between EmuSC and the real hardware depends on which instruments and effects that are used; some instruments are almost perfectly emulated, while others have notable differences.

I will update the readme file next time I work on it to reflect the actual improvements the last 6 moths.

PS: A way to check the current state of emulation is to compare the audio from emusc with some youtube videos of the same MIDI files played on the real hardware. This way you can hear the actual difference for those files.

**Reply by farvardin** [reply]
@skjelten thanks for the clarification, I was expecting a similar statement, because emusc sounds very good :)
I'll test and use it more for sure!
